docs/source/plugins/for_plugin_developers.rst

```diff
@@ -0,0 +1,274 @@
+   can implement.  If you have a plugin idea that requires napari to create a
+   new hook specification, we'd love to hear about it!  Please think about what
+   the signature of your proposed hook specification would look like, and where
+   within the napari codebase you'd like your hook implementation to be called,
+   and `open a feature request
+   <https://github.com/napari/napari/issues/new?template=feature_request.md>`_
+   in the napari issue tracker with your proposal.
+
+Let's take the :func:`~napari.plugins.hook_specifications.napari_get_reader`
+hook (our primary "reader plugin" hook) as an example.  It is defined as:
+
+.. code-block:: python
+
+   LayerData = Union[Tuple[Any], Tuple[Any, Dict], Tuple[Any, Dict, str]]
+   ReaderFunction = Callable[[str], List[LayerData]]
+
+   @napari_hook_specification(firstresult=True)
+   def napari_get_reader(
+       path: Union[str, List[str]]
+   ) -> Optional[ReaderFunction]:
+       ...
+
+Note that it takes a ``str`` or a ``list`` of ``str`` and either returns
+``None`` or a function.  From the docstring of the hook specification, we see
+that the implementation should return ``None`` if the path is of an
+unrecognized format, otherwise it should return a ``ReaderFunction``, which is
+a function that takes a ``str`` (the filepath to read) and returns a ``list``
+of ``LayerData``, where ``LayerData`` is any one of ``(data,)``, ``(data,
+meta)``, or ``(data, meta, layer_type)``.
+
+That seems like a bit of a mouthful!  But it's a precise (though flexible)
+contract that we can follow, and know that napari will handle the rest.
+
+
+Step 2: Write your hook implementation
+--------------------------------------
+
+Once you have identified the :ref:`hook specification <hook-specifications>`
+that you want to implement, you have to create a *hook implementation*: a
+function that accepts the arguments specified by the hook specification
+signature and returns a value with the expected return type.
+
+Here's an example hook implementation for
+:func:`~napari.plugins.hook_specifications.napari_get_reader` that enables
+napari to open an imaginary ``.ext`` filetype.
+
+.. code-block:: python
+
+   from pluggy import HookimplMarker
+
+   # we'll get to this line and the decorator below in just a minute
+   napari_hook_implementation = HookimplMarker("napari")
+
+   @napari_hook_implementation
+   def napari_get_reader(path):
+      # remember, path can be a list, so we check it's type first...
+      # (this example plugin doesn't handle lists)
+      if isinstance(path, str) and path.endswith(".ext"):
+         # If we recognize the format, we return the actual reader function
+         return my_reader
+      # otherwise we return None.
+      return None
+
+
+   def my_reader(path):
+      with open(path, 'rb') as file:
+         array = convert_bytes_to_numpy(data)
+      # return it as a list of LayerData
+      return [(array,)]
```

Comment:
I would move this definition above the `napari_get_reader` call and rename it `ext_file_reader`. To quote @stefanv from long ago, it's valid for the Python parser to define the function under the place where it's called, but it breaks my brain's parser. =P (He told *me* when I'd done the exact same thing. 😂)

Reply:
Also, I don't love that `convert_bytes_to_numpy` is undefined. Maybe just do a real plugin here for .npy files using np.load?